### ISSUE TITLE
QEMU

### ISSUE THREAD

**Issue by neuschaefer** [neuschaefer/wpcm450#10]
It would be interesting to have WPCM450 system emulation in QEMU. Some of the peripherals of NPCM7xx can probably be reused.

**Comment by Hurricos** [comment]
Based on [this FOSDEM presentation](https://habkost.net/talks/fosdem-2017/), there isn't any 'nice' way to translate a device tree directly into any of the QEMU device APIs, so a QEMU config would have to be built by hand. Still, I agree, it'd be very nice to verify booting works in QEMU before flashing the build directly e.g. to a motherboard.

> peripherals of NPCM7xx can probably be reused.

In what sense? Has a QEMU config already been set up for the NPCM7xx?

**Comment by neuschaefer** [comment]
QEMU already has peripheral device implementations and [SoC](https://github.com/qemu/qemu/blob/master/hw/arm/npcm7xx.c)/[board](https://github.com/qemu/qemu/blob/master/hw/arm/npcm7xx_boards.c) configs for NPCM750:

```
$ qemu-system-arm -M ? | grep NPCM
npcm750-evb          Nuvoton NPCM750 Evaluation Board (Cortex A9)
```

The configurations are probably too specific to the NPCM750(-evb) to reuse, but some of the device emulation code (e.g. [I2C](https://github.com/qemu/qemu/blob/266469947161aa10b1d36843580d369d5aa38589/hw/i2c/npcm7xx_smbus.c)) can probably be reused.


I may be misunderstanding what you mean by a "QEMU config". If that's the case, please elaborate.

**Comment by Hurricos** [comment]
No, you understood me correctly. I actually was not actually really cognizant that the Qemu Object Model actually gets used using C code. Thanks for the links to the SoC and boards.

So would a good way to build the device in QEMU be to loop through the following?:

1) take a minimal default QEMU arm32 SoC + Board,
2) rebuild Linux using select patches from your tree, where applicable
3) migrate an item described in the final device tree into both the compiled-in device tree and into the SoC / board config
4) rebuild qemu
5) Repeat from step 2

**Comment by neuschaefer** [comment]
Sounds like an approach that should work.

For an absolutely minimal configuration, I think you'll need:
- An interrupt controller (borrow from another platform, or implement the WPCM450 interrupt controller)
- A timer module (take the one from NPCM750, it's the same)
- A serial port (a generic 8250 might be good enough, or look at what the NPCM750 platform in QEMU does)
- set the CPU to ARM926EJ-S

If you have any questions regarding how a peripheral works, please use the wiki and/or raise issues with the "Documentation" tag. I'll be happy to help!

**Comment by StidOfficial** [comment]
I take versatile board like base board and take some code from ncpm7xx board, change base address and irq from your wiki.

Your 2 bare-metal programs (screem and monitor) your for UART0 output with serail0 on qemu but output not work.
I just configure UART address, so i think is important to configure interrupt controller and timer correctly for make monitor work correctly.

I continue to work on and publish as soon as possible the modification on github.

Thanks for you work

**Comment by neuschaefer** [comment]
Hi @StidOfficial,
The scream program only accesses the UART. I think it's a good first test case, because it is (almost) the simplest possible test program.
The monitor accesses the UART and the timer, but not the interrupt controller, so it should, at this point, not be necessary to hook up the interrupt controller.

**Comment by StidOfficial** [comment]
I have disable lot of code (SPI, GPIO and SMC) but we hare close for u-boot.

```
U-Boot 1.2.0-gbacd4830-dirty (Sep  8 2021 - 18:22:29) Avocent (1.13.7) Whoville

Board: WPCM450_00
DRAM:  128 MB
Flash:  0 kB
*** Warning - bad CRC, using default environment

In:    serial
Out:   serial
Err:   serial
Net:   No ethernet found.
Hit any key to stop autoboot:  0 
[uboot_wpcm450]# 
```

**Comment by StidOfficial** [comment]
@neuschaefer Do you check https://datasheetspdf.com/pdf-file/1134405/nuvoton/NUC960ADN/1 beacause there are very somilar registers, base addresses, ... ?

**Comment by neuschaefer** [comment]
@StidOfficial thanks for the link! :)
